util/fibers/synchronization.h

```diff
@@ -404,16 +406,28 @@ class BlockingCounter {
       }
     }
 
+    auto WaitCondition(uint64_t* cnt) {
+      return [this, cnt]() -> bool {
+        *cnt = count_.load(std::memory_order_acquire);
+        return *cnt == 0 || (*cnt & kCancelFlag);
+      };
+    }
+
     // I suspect all memory order accesses here could be "relaxed" but I do not bother.
-    void Wait() {
-      ec_.await([this] {
-        auto cnt = count_.load(std::memory_order_acquire);
-        return cnt == 0 || (cnt & (1ULL << 63));
-      });
+    bool Wait() {
+      uint64_t cnt;
+      ec_.await(WaitCondition(&cnt));
+      return (cnt & kCancelFlag) == 0;
+    }
+
+    bool WaitUntil(const std::chrono::steady_clock::time_point& tp) {
+      uint64_t cnt;
+      std::cv_status status = ec_.await_until(WaitCondition(&cnt), tp);
+      return status == std::cv_status::no_timeout && (cnt & kCancelFlag) == 0;
```

Comment:
```suggestion
      return status != std::cv_status::no_timeout && (cnt & kCancelFlag) == 0;
```

Reply:
Why? Success = no timeout reached and not cancelled

Reply:
sorry my mistake here I read timeout instead NO timeout
please ignore

Reply:
Please re-review when you have time 🙂